README
NAME

Time::TT - Terrestrial Time and its realisations

DESCRIPTION

Terrestrial Time (TT) is a time scale representing time on the surface
of Terra.  Specifically, it is the proper time experienced by a clock
located on the rotating geoid (i.e., at sea level).  It is indirectly the
basis for Terran civil timekeeping, via its realisation International
Atomic Time (TAI).  It is linearly related to (and in fact now defined
in terms of) the astronomical time scale Geocentric Coordinate Time (TCG).

This module represents instants on the TT time scale as a scalar number
of SI seconds since an epoch.  This is an appropriate form for all manner
of calculations.  The TT scale is defined with a well-known point at
TAI instant 1977-01-01T00:00:00.0.  That instant is assigned the scalar
value 599616000 exactly, corresponding to an epoch (scalar value zero)
near the TAI epoch 1958-01-01T00:00:00.0.  This matches the convention
used by "Time::TAI" for instants on the TAI scale.  Because TAI does not
match the rate of TT perfectly, the TT epoch is not precisely equal to
the TAI epoch, but is instead around 600 us earlier than it.

There is also a conventional way to represent TT instants using day-based
notations associated with planetary rotation `time' scales.  The `day'
of TT is a nominal period of exactly 86400 SI seconds, which is slightly
shorter than an actual Terran day.  The well-known point at TAI instant
1977-01-01T00:00:00.0 is assigned the label 1977-01-01T00:00:32.184
(MJD 43144.0003725).  Because TT is not connected to Terran rotation,
and so has no inherent concept of a day, it is somewhat misleading to
use such day-based notations.  Conversion between this notation and the
linear count of seconds is supported by this module.  This notation does
not match the similar day-based notation used for TAI.

There is another conventional way to represent TT instants, using a larger
unit approximating the duration of a Terran year.  The `Julian year' 
is a nominal period of exactly 365.25 `days' of exactly 86400 SI seconds
each.  The TT instant 2000-01-01T12:00:00.0 (MJD 51544.5) is labelled as
Julian epoch 2000.0.  Julian epochs are used only with TT, not with any
other time scale.  The Julian epoch numbers correspond approximately to
Gregorian calendar years, for dates within a few kiloyears of the epoch.
Because TT is not connected to the Terran orbit, and so has no inherent
concept of a year, the year-based notation is somewhat misleading.
Conversion between this notation and the linear count of seconds is
supported by this module.

Because TT is a theoretical time scale, not directly accessible for
practical use, it must be realised using atomic clocks.  This is done
by metrological agencies, each with different imperfections.  To achieve
microsecond accuracy it is necessary to take account of these differences.
This module supports conversion of times between different realisations
of TT.

INSTALLATION

	perl Build.PL
	./Build
	./Build test
	./Build install

AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

COPYRIGHT

Copyright (C) 2006, 2007, 2010 Andrew Main (Zefram) <zefram@fysh.org>

LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
version 0.004; 2010-07-22

  * bugfix: require bugfixed versions of Date::JD and Time::UTC (for
    $SIG{__DIE__} handling)

  * in Time::TT::OffsetKnot, be more stringent in parsing supplied data

  * in Time::TT::tt_realisation(), more cleanly reject "bipmNN"
    realisation names that include non-ASCII digits

  * be more stringent in parsing time scale data files

  * regularise LWP version checking code

  * use simpler "parent" pragma in place of "base"

  * in Time::TT::Agencies documentation, note new agencies BY, EIM,
    INTI, KIM, KZ, LV, NRL, SIQ, SMD, UA, VMI

  * in Time::TT::Agencies documentation, note renamings NMC->BIM,
    ZMDM->DMDM, IGMA->IGNA, OMH->MKEH, CSIR->ZA

  * document standard Time::TT::Realisation methods in
    Time::TT::InterpolatingRealisation

  * document standard Math::Interpolator::Knot methods in
    Time::TT::OffsetKnot

  * in documentation, use "^" instead of "**" to signify exponentiation

  * in Time::TT::OffsetKnot, abandon use of the "fields" module

  * POD markup fix

  * use full stricture in test suite

  * test POD syntax and coverage, and rename some internal functions to
    satisfy the coverage test

  * drop prototypes from method subs (where the prototypes have no effect)

  * build with Module::Build instead of ExtUtils::MakeMaker

  * complete dependency list

  * include signature in distribution

  * in documentation, separate "license" section from "copyright" section

  * check for required Perl version at runtime

version 0.003; 2007-02-05

  * new functions tt_instant_to_jepoch() and tt_jepoch_to_instant()
    to handle interconversion with Julian epochs

  * in documentation, reference Time::TCG

  * test explicitly for correct class of result

version 0.002; 2007-01-21

  * in Time::TT::OffsetKnot, avoid "my __PACKAGE__", for compatibility
    with perl v5.6

  * typo fix in Time::TT::Agencies documentation

version 0.001; 2006-08-03

  * list exact required modules in Makefile.PL

  * use "=> 0" instead of "=> undef" in unversioned dependencies in
    Makefile.PL

  * don't include underscore in string version of module version numbers
    in Makefile.PL

  * use "warnings" and "strict" pragmas in all modules

  * documentation clarifications

version 0.000; 2006-06-13

  * initial released version


Build.PL
{ use 5.006; }
use warnings;
use strict;

use Module::Build;

Module::Build->new(
	module_name => "Time::TT",
	license => "perl",
	configure_requires => {
		"Module::Build" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	build_requires => {
		"Math::BigRat" => "0.04",
		"Module::Build" => 0,
		"Test::More" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	requires => {
		"Carp" => 0,
		"Date::JD" => "0.002",
		"Exporter" => 0,
		"LWP" => "5.5394",
		"LWP::UserAgent" => 0,
		"Math::BigRat" => "0.04",
		"Math::Interpolator::Knot" => 0,
		"Math::Interpolator::Linear" => 0,
		"Math::Interpolator::Robust" => 0,
		"Math::Interpolator::Source" => 0,
		"Time::UTC" => "0.005",
		"constant" => 0,
		"parent" => 0,
		"perl" => "5.006",
		"strict" => 0,
		"warnings" => 0,
	},
	dynamic_config => 0,
	meta_add => { distribution_type => "module" },
	create_makefile_pl => "passthrough",
	sign => 1,
)->create_build_script;

1;


lib/Time/TT/OffsetKnot.pm
=head1 NAME

Time::TT::OffsetKnot - MJD/offset tuple for realisations of TT

=head1 SYNOPSIS

	use Time::TT::OffsetKnot;

	$pt = Time::TT::OffsetKnot($mjd, $offset_ns, 9);

	$tai_instant = $pt->x;
	$instant = $pt->y;
	$role = $pt->role;

=head1 DESCRIPTION

This class assists in implementing a realisation of Terrestrial Time
(TT) by interpolation using the C<Time::TT::InterpolatingRealisation>
class.  Many such time scales are defined by tables of data giving the
difference between TAI and the time scale at 00:00 UTC on various days.
An object of this class represents such a data point (knot), supplying
the C<Math::Interpolator::Knot> interface required.

The calculations required in order to generate the actual numbers to
interpolate with are performed lazily and memoised.  Until these numbers
(which will be C<Math::BigRat> objects) are generated, the raw data is
stored very space-efficiently.  Construction of the knot object is quick.

=cut

package Time::TT::OffsetKnot;

{ use 5.006; }
use warnings;
use strict;

use Carp qw(croak);
use Date::JD 0.002 qw(mjd_to_cjdn);
use Math::BigRat 0.04;
use Time::UTC 0.005 qw(utc_to_tai utc_cjdn_to_day);

our $VERSION = "0.004";

=head1 CONSTRUCTOR

=over

=item Time::TT::OffsetKnot->new(MJD, OFFSET, UNIT)

Creates and returns an object representing a knot where at the integral
Modified Julian Date MJD (interpreted according to UTC) the time scale of
interest differed from TAI by OFFSET * 10^-UNIT seconds.  For example,
if OFFSET is 43.5 and UNIT is 9 then the time scale differs from TAI by
43.5 ns.

MJD must be purely a string of digits.  OFFSET must be a string
in decimal numeric syntax, matching C</[-+]?[0-9]+(?:\.[0-9]+)?/>.
UNIT must be non-negative.

If the OFFSET value is positive then it indicates that the time scale
is behind TAI.  If the OFFSET value is negative then the time scale is
ahead of TAI.  This is the sense of the offset found in TAI-TAI(k) files.
If the raw data is in the sense TT(k)-TT(TAI) then the sign will have
to be reversed before passing it to this constructor.

=cut

sub new {
	my($class, $mjd, $offset, $unit) = @_;
	croak "malformed MJD `$mjd'" unless $mjd =~ /\A[0-9]+\z/;
	my($sign, $int, $frac) =
		($offset =~ /\A([-+]?)([0-9]+)(?:\.([0-9]+))?\z/);
	croak "malformed offset `$offset'" unless defined $int;
	$frac = "" unless defined $frac;
	$int = ("0" x $unit).$int;
	my $pos = length($int) - $unit;
	return bless({
		x => $mjd,
		y => $sign.substr($int, 0, $pos).".".substr($int, $pos).$frac,
	}, $class);
}

=back

=head1 METHODS

=over

=item $pt->x

=item $pt->y

=item $pt->role

These methods are part of the standard C<Math::Interpolator::Knot>
interface.

=cut

use constant BIGRAT_ZERO => Math::BigRat->new(0);

sub x {
	my($self) = @_;
	my $x = $self->{x};
	if(ref($x) eq "") {
		$x = utc_to_tai(
			utc_cjdn_to_day(mjd_to_cjdn(
				Math::BigRat->new($x), BIGRAT_ZERO)),
			BIGRAT_ZERO);
		$self->{x} = $x;
	}
	return $x;
}

sub y {
	my($self) = @_;
	my $y = $self->{y};
	if(ref($y) eq "") {
		$y = $self->x - Math::BigRat->new($y);
		$self->{y} = $y;
	}
	return $y;
}

sub role { "KNOT" }

=back

=head1 SEE ALSO

L<Math::Interpolator::Knot>,
L<Time::TT::InterpolatingRealisation>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2006, 2007, 2010 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/Time/TT/Realisation.pm
=head1 NAME

Time::TT::Realisation - realisation of Terrestrial Time (base class)

=head1 SYNOPSIS

	$tai_instant = $rln->to_tai($instant);
	$instant = $rln->from_tai($tai_instant);

	$rln1_instant = $rln0->to_realisation($rln1, $rln0_instant);
	$rln0_instant = $rln0->from_realisation($rln1, $rln1_instant);

=head1 DESCRIPTION

Terrestrial Time (TT) is a platonic time scale, consisting of the
time axis of a relativistic reference frame located on the Terran
geoid.  It is not directly accessible for practical use.  Instead,
it must be realised by the use of atomic clocks and other equipment.
A realisation is inevitably imperfect, only approximating the true TT.
Consequently there are many realisations, with differing degrees of
accuracy and accessibility.

An object of this class represents a realisation of TT.  The main
use of such an object is to convert between realisations in order to
determine the time of events with very high precision, in cases where
the directly-accessible realisation is not sufficiently accurate.

This is a base class, defining the interface for realisation objects.
If you already have a realisation object then this document is what
you should read in order to know how to use it.  If you don't have
a realisation object yet, there are no constructors here: see the
C<tt_realisation> function in L<Time::TT> or the C<tai_realisation>
function in L<Time::TAI>.  If you are implementing a subclass of
realisation object, see L</SUBCLASSING>.

The principal realisation of TT is International Atomic Time (TAI).
This is defined retrospectively, in monthly bulletins from the BIPM,
by its relation to real-time approximations of TAI that are supplied in
public time signals by tens of metrological agencies around the world.
Better realisations of TT are defined further in retrospect, and are
defined by their relation to TAI.  TAI thus has a pivotal role: different
realisations of TT can be related to each other by using TT(TAI) as an
intermediate form.

In this interface, instants on the TT scale are represented as a scalar
number of seconds since the TT epoch, as described in L<Time::TT>.
All such numbers are represented as C<Math::BigRat> objects.

=cut

package Time::TT::Realisation;

{ use 5.006; }
use warnings;
use strict;

our $VERSION = "0.004";

=head1 METHODS

=over

=item $rln->to_tai(INSTANT)

Takes an instant expressed on the time scale represented by this object,
and converts it to an instant on the TT(TAI) scale.  The input must be
a C<Math::BigRat> object, and the result is the same type.

=item $rln->from_tai(TAI_INSTANT)

Takes an instant expressed on the TT(TAI) scale, and converts it to an
instant on the time scale represented by this object.  The input must
be a C<Math::BigRat> object, and the result is the same type.

=item $rln->to_realisation(REALISATION, INSTANT)

Takes an instant expressed on the time scale represented by this object,
and converts it to an instant on the time scale represented by the
REALISATION object.  The input must be a C<Math::BigRat> object, and
the result is the same type.

=cut

sub to_realisation {
	my($self, $rln, $t) = @_;
	return $t if $rln == $self;
	return $rln->from_tai($self->to_tai($t));
}

=item $rln->from_realisation(REALISATION, INSTANT)

Takes an instant expressed on the time scale represented by the
REALISATION object, and converts it to an instant on the time scale
represented by this object.  The input must be a C<Math::BigRat> object,
and the result is the same type.

=cut

sub from_realisation {
	my($self, $rln, $t) = @_;
	return $t if $rln == $self;
	return $self->from_tai($rln->to_tai($t));
}

=back

=head1 SUBCLASSING

This class is designed to be subclassed, and cannot be instantiated alone.
Any subclass must implement the C<to_tai> and C<from_tai> methods.
That is the minimum required.  The general C<to_realisation> and
C<from_realisation> methods may also be implemented for cases that can
be done efficiently; they have default implementations in this class in
terms of C<to_tai> and C<from_tai>.

=head1 SEE ALSO

L<Time::TAI>,
L<Time::TT>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2006, 2007, 2010 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/Time/TT/InterpolatingRealisation.pm
=head1 NAME

Time::TT::InterpolatingRealisation - TT realised by interpolation

=head1 SYNOPSIS

	use Time::TT::InterpolatingRealisation;

	$rln = Time::TT::InterpolatingRealisation->new($interpolator);

	$tai_instant = $rln->to_tai($instant);
	$instant = $rln->from_tai($tai_instant);
	$rln1_instant = $rln0->to_realisation($rln1, $rln0_instant);
	$rln0_instant = $rln0->from_realisation($rln1, $rln1_instant);

=head1 DESCRIPTION

This class implements a realisation of Terrestrial Time (TT) by
interpolation between known points of correlation between the realisation
and International Atomic Time (TAI).  See L<Time::TT::Realisation>
for the interface.

=cut

package Time::TT::InterpolatingRealisation;

{ use 5.006; }
use warnings;
use strict;

our $VERSION = "0.004";

use parent "Time::TT::Realisation";

=head1 CONSTRUCTOR

Normally one won't use this constructor directly.  See the
C<tt_realisation> function in L<Time::TT>, which will construct a range
of published realisations, most of which are implemented using this class.
Use this directly only if the realisation that you desire is not available
by that means.

=over

=item Time::TT::InterpolatingRealisation->new(INTERPOLATOR)

Constructs and returns an object representing a realisation of
TT that is defined by isolated points of correlation between it
and TAI.  The INTERPOLATOR argument must be an object of a subclass
of C<Math::Interpolator>, supplying the C<x> and C<y> methods.  The x
coordinate of the interpolator's curve must represent TAI, and the y
coordinate the realisation of interest.  Times on both coordinates are
represented as the number of seconds since the 1958 epoch, as described
in L<Time::TT>.  All numbers must be C<Math::BigRat> objects.

The class C<Time::TT::OffsetKnot> may be useful in building the required
interpolator.

=cut

sub new {
	my($class, $interpolator) = @_;
	return bless(\$interpolator, $class);
}

=back

=head1 METHODS

=over

=item $rln->to_tai(INSTANT)

=item $rln->from_tai(TAI_INSTANT)

These methods are part of the standard C<Time::TT::Realisation> interface.

=cut

sub to_tai {
	my($self, $t) = @_;
	return ${$self}->x($t);
}

sub from_tai {
	my($self, $t) = @_;
	return ${$self}->y($t);
}

=back

=head1 SEE ALSO

L<Math::BigRat>,
L<Math::Interpolator>,
L<Time::TT>,
L<Time::TT::OffsetKnot>,
L<Time::TT::Realisation>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2006, 2007, 2010 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;


lib/Time/TT.pm
=head1 NAME

Time::TT - Terrestrial Time and its realisations

=head1 SYNOPSIS

	use Time::TT qw(tt_instant_to_mjd tt_mjd_to_instant);

	$mjd = tt_instant_to_mjd($instant);
	$instant = tt_mjd_to_instant($mjd);

	use Time::TT qw(tt_instant_to_jepoch tt_jepoch_to_instant);

	$jepoch = tt_instant_to_jepoch($instant);
	$instant = tt_jepoch_to_instant($jepoch);

	use Time::TT qw(tt_realisation);

	$rln = tt_realisation("bipm05");
	$instant = $rln->from_tai($tai_instant);

=head1 DESCRIPTION

Terrestrial Time (TT) is a time scale representing time on the surface
of Terra.  Specifically, it is the proper time experienced by a clock
located on the rotating geoid (i.e., at sea level).  It is indirectly the
basis for Terran civil timekeeping, via its realisation International
Atomic Time (TAI).  It is linearly related to (and in fact now defined
in terms of) the astronomical time scale Geocentric Coordinate Time (TCG).

This module represents instants on the TT time scale as a scalar number
of SI seconds since an epoch.  This is an appropriate form for all manner
of calculations.  The TT scale is defined with a well-known point at
TAI instant 1977-01-01T00:00:00.0.  That instant is assigned the scalar
value 599_616_000 exactly, corresponding to an epoch (scalar value zero)
near the TAI epoch 1958-01-01T00:00:00.0.  This matches the convention
used by C<Time::TAI> for instants on the TAI scale.  Because TAI does
not match the rate of TT perfectly, the TT epoch is not precisely equal
to the TAI epoch, but is instead around 600 us earlier than it.

There is also a conventional way to represent TT instants using day-based
notations associated with planetary rotation `time' scales.  The `day'
of TT is a nominal period of exactly 86400 SI seconds, which is slightly
shorter than an actual Terran day.  The well-known point at TAI instant
1977-01-01T00:00:00.0 is assigned the label 1977-01-01T00:00:32.184
(MJD 43144.0003725).  Because TT is not connected to Terran rotation,
and so has no inherent concept of a day, it is somewhat misleading to
use such day-based notations.  Conversion between this notation and the
linear count of seconds is supported by this module.  This notation does
not match the similar day-based notation used for TAI.

There is another conventional way to represent TT instants, using a larger
unit approximating the duration of a Terran year.  The `Julian year'
is a nominal period of exactly 365.25 `days' of exactly 86400 SI seconds
each.  The TT instant 2000-01-01T12:00:00.0 (MJD 51544.5) is labelled as
Julian epoch 2000.0.  Julian epochs are used only with TT, not with any
other time scale.  The Julian epoch numbers correspond approximately to
Gregorian calendar years, for dates within a few kiloyears of the epoch.
Because TT is not connected to the Terran orbit, and so has no inherent
concept of a year, the year-based notation is somewhat misleading.
Conversion between this notation and the linear count of seconds is
supported by this module.

Because TT is a theoretical time scale, not directly accessible for
practical use, it must be realised using atomic clocks.  This is done
by metrological agencies, each with different imperfections.  To achieve
microsecond accuracy it is necessary to take account of these differences.
This module supports conversion of times between different realisations
of TT.

=cut

package Time::TT;

{ use 5.006; }
use warnings;
use strict;

use Carp qw(croak);
use Math::BigRat 0.04;

our $VERSION = "0.004";

use parent "Exporter";
our @EXPORT_OK = qw(
	tt_instant_to_mjd tt_mjd_to_instant
	tt_instant_to_jepoch tt_jepoch_to_instant
	tt_realisation
);

=head1 FUNCTIONS

=over

=item tt_instant_to_mjd(INSTANT)

Converts from a count of seconds to a Modified Julian Date in the manner
conventional for TT.  The MJD can be further converted to other forms of
day-based date using other modules.  The input must be a C<Math::BigRat>
object, and the result is the same type.

=cut

use constant TT_EPOCH_MJD => Math::BigRat->new("36204.0003725");

sub tt_instant_to_mjd($) {
	my($tt) = @_;
	return TT_EPOCH_MJD + ($tt / 86400);
}

=item tt_mjd_to_instant(MJD)

Converts from a Modified Julian Date, interpreted in the manner
conventional for TT, to a count of seconds.  The input must be a
C<Math::BigRat> object, and the result is the same type.

=cut

sub tt_mjd_to_instant($) {
	my($mjd) = @_;
	return ($mjd - TT_EPOCH_MJD) * 86400;
}

=item tt_instant_to_jepoch(INSTANT)

Converts from a count of seconds to a Julian epoch.  The input must be
a C<Math::BigRat> object, and the result is the same type.

=cut

use constant TT_EPOCH_JEPOCH => 1958 + Math::BigRat->new("0.0003725/365.25");

sub tt_instant_to_jepoch($) {
	my($tt) = @_;
	return TT_EPOCH_JEPOCH + ($tt / 31557600);
}

=item tt_jepoch_to_instant(JEPOCH)

Converts from a Julian epoch to a count of seconds.  The input must be
a C<Math::BigRat> object, and the result is the same type.

=cut

sub tt_jepoch_to_instant($) {
	my($jepoch) = @_;
	return ($jepoch - TT_EPOCH_JEPOCH) * 31557600;
}

=item tt_realisation(NAME)

Looks up and returns an object representing a named realisation of TT.
The object returned is of the class C<Time::TT::Realisation>; see the
documentation of that class for its interface.

The name, recognised case-insensitively, may be of these forms:

=over

=item B<bipm05>

Retrospective best estimate of TT, published by the BIPM.  TT(BIPM05)
was published in 2005, and other versions were (and will be) published
in other years, with the digits in the name varying accordingly.
These time scales are currently based on reanalysis of the TAI data.
They are defined by isolated data points, so conversions in general
involve interpolation; the process is by its nature inexact.

=item B<eal>

TT(EAL) is derived from the Free Atomic Scale (EAL).  EAL is the
weighted average of the time ticked by the clocks contributing to TAI,
with no gravitational correction applied.  TAI is generated by applying
a frequency shift to EAL to correct for gravitational time dilation.
The relationship between EAL and TAI is precisely defined, so conversions
are exact.

=item B<tai>

TT(TAI) is the principal realisation of TT, derived directly from
International Atomic Time (TAI).  This is defined monthly in retrospect
and then never revised.

=item B<tai/npl>

TT(TAI) based on TAI(NPL), the real-time estimate of TAI supplied by
the National Physical Laboratory in the UK.  Other real-time estimates
of TAI are named similarly using an abbreviation of the name of the
supplying agency.  See the C<tai_realisation> function in L<Time::TAI>
for more discussion, or L<Time::TT::Agencies> for a list of agencies.

=back

Other names may be recognised in the future, as more TT(k) time scales
are defined.

In order to use any of the TAI-based realisations the C<Time::TAI>
module is required.

=cut

#
# general
#

sub _get_bipm_file($) {
	my($fn) = @_;
	require LWP;
	LWP->VERSION(5.53_94);
	require LWP::UserAgent;
	my $response = LWP::UserAgent->new
				->get("ftp://ftp2.bipm.fr/pub/tai/$fn");
	croak "can't download $fn: ".$response->message
		unless $response->code == 200;
	return $response->content;
}

#
# TT(BIPMnn)
#

use constant TT_SYNCH_TIME => Math::BigRat->new(599616000);
use constant TT_SYNCH_MJD => 43144;

my %tt_bipmnn;

sub _tt_bipmnn($) {
	my($yr) = @_;
	my $r = $tt_bipmnn{$yr};
	return $r if defined $r;
	my $content = _get_bipm_file("scale/ttbipm.$yr");
	$content =~ /\A[\ \t\n]*TT\(BIPM[0-9]{2,4}\)\ is\ a\ realization/
		or die "doesn't look like a TT(BIPMnn) file\n";
	require Time::TT::OffsetKnot;
	my @data;
	my $last_mjd = 0;
	while($content =~ /^\ *([0-9]+)\.(?:[-+]|\ +[-+]?)[0-9]+(?:\.[0-9]+)?
			   ([-+]|\ +[-+]?)([0-9]+(?:\.[0-9]+)?)\ *[\r\n]/xmg) {
		my($mjd, $sign, $offset_us) = ($1, $2, $3);
		die "data out of order at mjd=$mjd" unless $mjd > $last_mjd;
		if($last_mjd < TT_SYNCH_MJD && $mjd >= TT_SYNCH_MJD) {
			require Math::Interpolator::Knot;
			push @data, Math::Interpolator::Knot
					->new(TT_SYNCH_TIME, TT_SYNCH_TIME);
		}
		$offset_us = "-$offset_us" unless $sign =~ /-\z/;
		push @data, Time::TT::OffsetKnot->new($mjd, $offset_us, 6);
		$last_mjd = $mjd;
	}
	require Math::Interpolator::Robust;
	require Time::TT::InterpolatingRealisation;
	$r = Time::TT::InterpolatingRealisation->new(
		Math::Interpolator::Robust->new(@data));
	return $tt_bipmnn{$yr} = $r;
}

#
# TT(EAL)
#

my $tt_eal;

sub _tt_eal() {
	return $tt_eal if defined $tt_eal;
	my $content = _get_bipm_file("scale/ealtai04.ar");
	$content =~ /\A[\ \t\n]*[^\n]*differences between the normalized/i
		or die "doesn't look like an EAL file\n";
	require Math::Interpolator::Knot;
	my @data;
	my $tai = Math::BigRat->new(-94694400);   # 1955-01-01
	push @data, Math::Interpolator::Knot->new($tai, $tai);
	$tai = Math::BigRat->new(599616000);   # 1977-01-01
	push @data, Math::Interpolator::Knot->new($tai, $tai);
	my $mjd = Math::BigRat->new(43144);
	my $eal = $tai;
	my $fdiff_scale = Math::BigRat->new("0.0000000000001");
	while($content =~ /^\ *[0-9]+\ +[A-Za-z]+\ +[0-9]+\ +-
			    \ +[0-9]+\ +[A-Za-z]+\ +[0-9]+
			    \ +([0-9]+)\ +-\ +([0-9]+)
			    \ +([0-9]+(?:\.[0-9]+)?)[\ \t\n]/xmg) {
		my($old_mjd, $new_mjd, $fdiff) = ($1, $2, $3);
		$old_mjd = Math::BigRat->new($old_mjd);
		die "data not contiguous at mjd=$mjd" unless $old_mjd == $mjd;
		$new_mjd = Math::BigRat->new($new_mjd);
		$fdiff = Math::BigRat->new($fdiff) * $fdiff_scale;
		my $tai_s = ($new_mjd - $mjd) * 86400;
		my $eal_s = $tai_s * (1 + $fdiff);
		$mjd = $new_mjd;
		$tai += $tai_s;
		$eal += $eal_s;
		push @data, Math::Interpolator::Knot->new($tai, $eal);
	}
	$tai += 1000000;
	$eal += 1000000;
	require Math::Interpolator::Source;
	push @data, Math::Interpolator::Source->new(
			sub () { croak "later data for TT(EAL) is missing"; },
			$tai, $eal);
	require Math::Interpolator::Linear;
	require Time::TT::InterpolatingRealisation;
	$tt_eal = Time::TT::InterpolatingRealisation->new(
			Math::Interpolator::Linear->new(@data));
	return $tt_eal;
}

#
# invocation of realisations
#

sub tt_realisation($) {
	my($k) = @_;
	if($k =~ m#\Atai(?:/(.+))?\z#si) {
		require Time::TAI;
		return Time::TAI::tai_realisation(defined($1) ? $1 : "");
	} elsif($k =~ m#\Abipm([0-9][0-9])\z#i) {
		return _tt_bipmnn($1);
	} elsif($k =~ m#\Aeal\z#i) {
		return _tt_eal();
	} else {
		croak "no realisation TT(".uc($k).") known";
	}
}

=back

=head1 BUGS

The data for EAL only goes forward to mid-2005.  There is no
machine-readable source of subsequent data.

=head1 SEE ALSO

L<Date::JD>,
L<Time::TAI>,
L<Time::TCG>,
L<Time::TT::Agencies>,
L<Time::TT::Realisation>

=head1 AUTHOR

Andrew Main (Zefram) <zefram@fysh.org>

=head1 COPYRIGHT

Copyright (C) 2006, 2007, 2010 Andrew Main (Zefram) <zefram@fysh.org>

=head1 LICENSE

This module is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1;
